bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/handler/NanoleafControllerHandler.java

```diff
@@ -129,41 +148,43 @@ public void initialize() {
         setPort(config.port);
         setRefreshIntervall(config.refreshInterval);
         setAuthToken(config.authToken);
-
         Map<String, String> properties = getThing().getProperties();
         String propertyModelId = properties.get(Thing.PROPERTY_MODEL_ID);
         if (hasTouchSupport(propertyModelId)) {
             config.deviceType = DEVICE_TYPE_TOUCHSUPPORT;
+            initializeTouchHttpClient();
```

Comment:
Like so?

            scheduler.execute(() -> {
                initializeTouchHttpClient();
            });
            
An issue I am seeing here is that in line 184 I am starting the touchJob which requires that the touchhttpClient to be available.  If it wasn't already at that  point (as it might not yet available from the scheduled task) I had to also schedule that in a repetitive way until this is case and cancel that scheduler when that is finally successfully. (though don't really have an idea how to do that in a reliable way without creating more side effects).

Any recommendations on a good and maintainable pattern to achieve this?

Reply:
Exactly!

You could move the entire rest of the initialization into the same block, this way everything is still executed in sequence, but the initialize() method itself returns quickly.

Reply:
Let me come back to the line 184 with startTouchJob() as I am struggling to find a good approach:

![image](https://user-images.githubusercontent.com/5937600/131842568-8d829b9c-8f36-4ca2-bd9f-3f5de3c2a7fd.png)

As you can see it depends on 164 (it can only start the touch job if the http client has been initialized). However, 184->startTouchJob() is deeply buried in further conditions and would be hard to be refactored to be put right next to or into the execute-lambda. 

The only idea I could think of is to wrap startTouchJob ( sc into another scheduler that checks its prerequisites and reschedules itself until they are met which may be the cleanest solution but maybe over-engineered?

Would you go that way /  far?

Reply:
I would put everything except the basic initialization (i.e. setting variables and fields in the `scheduler.execute` block:
```java
    public void initialize() {
        logger.debug("Initializing the controller (bridge)");
        updateStatus(ThingStatus.UNKNOWN);
        NanoleafControllerConfig config = getConfigAs(NanoleafControllerConfig.class);
        setAddress(config.address);
        setPort(config.port);
        setRefreshIntervall(config.refreshInterval);
        setAuthToken(config.authToken);

        scheduler.execute(() -> {
        Map<String, String> properties = getThing().getProperties();
        String propertyModelId = properties.get(Thing.PROPERTY_MODEL_ID);
        if (hasTouchSupport(propertyModelId)) {
            config.deviceType = DEVICE_TYPE_TOUCHSUPPORT;
            initializeTouchHttpClient();
        } else {
            config.deviceType = DEVICE_TYPE_LIGHTPANELS;
        }
        setDeviceType(config.deviceType);

        String propertyFirmwareVersion = properties.get(Thing.PROPERTY_FIRMWARE_VERSION);

        try {
            if (!config.address.isEmpty() && !String.valueOf(config.port).isEmpty()) {
                if (propertyFirmwareVersion != null && !propertyFirmwareVersion.isEmpty() && !OpenAPIUtils
                        .checkRequiredFirmware(properties.get(Thing.PROPERTY_MODEL_ID), propertyFirmwareVersion)) {
                    logger.warn("Nanoleaf controller firmware is too old: {}. Must be equal or higher than {}",
                            propertyFirmwareVersion, API_MIN_FW_VER_LIGHTPANELS);
                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
                            "@text/error.nanoleaf.controller.incompatibleFirmware");
                    stopAllJobs();
                } else if (config.authToken != null && !config.authToken.isEmpty()) {
                    stopPairingJob();
                    startUpdateJob();
                    startTouchJob();
                } else {
                    logger.debug("No token found. Start pairing background job");
                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_PENDING,
                            "@text/error.nanoleaf.controller.noToken");
                    startPairingJob();
                    stopUpdateJob();
                }
            } else {
                logger.warn("No IP address and port configured for the Nanoleaf controller");
                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_PENDING,
                        "@text/error.nanoleaf.controller.noIp");
                stopAllJobs();
            }
        } catch (IllegalArgumentException iae) {
            logger.warn("Nanoleaf controller firmware version not in format x.y.z: {}",
                    getThing().getProperties().get(Thing.PROPERTY_FIRMWARE_VERSION));
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
                    "@text/error.nanoleaf.controller.incompatibleFirmware");
        }
        }
    }
```
Could probably be restructured better, but something like this

Reply:
OK, you basically just defer everything into the thread. Let me research deeper if that might have some side effect. Thanks for your time, Anders, so far. I really appreciate that.